### PR TITLE
#6407 - Document Borg 1.2 pattern behavior change

### LINKED ISSUE #6407: borg 1.2 regular expression pattern behavior change
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Seems like a bug as it's not announced behaviour change, and dangerous as it suddenly includes files (or no longer includes files) it has included befor.

#### Your borg version (borg -V).

1.2 locally, no idea remote

#### Operating system (distribution) and version.

Ubuntu 22.04

#### Hardware / network configuration, and filesystems used.

-

#### How much data is handled by borg?

-

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

`borg create --progress --one-file-system --exclude-caches --info --stats --exclude-from=/home/jak/.config/borg/exclude.conf -C zlib,9 $repostory::$archive /home/jak`

## Describe the problem you're observing.

`re:` patterns no longer apply to the full filename, but to the relative one in the archive. Meaning I suddenly had all my regex excluded files pop up in my backups, such as files in `/home/jak/mail/.notmuch` which are huge. I had to remove the leading `/` to get the pattern to match again, as can be seen in https://github.com/julian-klode/dotfiles/commit/7a6f1e0437fe96311c59a38cbf2c743043941e31 and in this excerpt:

```diff
-re:/home/jak/.*/\.notmuch
+re:home/jak/.*/\.notmuch
```

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

N/A

### PR DESCRIPTION
**Pull request by fantasya-pbem**
Remove leading slashes in examples. Replace paragraph describing pattern behaviour change in 1.2.

fixes #6407

### CODE REVIEW

File: docs/usage/help.rst.inc [hunk 1]
```diff
@@ -31,10 +31,11 @@ path prefixes and path full-matches. By default, fnmatch is used for
...
 
-Starting with Borg 1.2, for all but regular expression pattern matching
-styles, all paths are treated as relative, meaning that a leading path
-separator is removed after normalizing and before matching. This allows
-you to use absolute or relative patterns arbitrarily.
+Starting with Borg 1.2, discovered fs paths are normalised, have leading
+slashes removed and then are matched against your patterns.
+Note: You need to review your include / exclude patterns and make
+sure they do not expect leading slashes. Borg can only deal with this
+for some very simple patterns by removing leading slashes there also.
```
Comment (ThomasWaldmann) [inline docs/usage/help.rst.inc:38]: see line 1 of this file. needs to be changed at the source of this, archiver.py?

File: docs/changes.rst [hunk 2]
```diff
@@ -305,6 +305,10 @@ and maybe just were not noticed.
 
 Compatibility notes:
 
+- matching of path patterns has been aligned with borg storing relative paths.
+  Borg archives file paths without leading slashes. Previously, include/exclude
+  patterns could contain leading slashes. You should check your patterns and
+  remove leading slashes.
```
Comment (ThomasWaldmann) [inline docs/changes.rst:311]: good, I'll copy that to above 1.2.1 entry when releasing that, so people will hopefully read it.

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]

**Comment by ThomasWaldmann** [comment]
maybe also proofread `helptext['patterns']` in `archiver.py` again if it correctly describes the relative path matching behaviour. 

do a global search for `--exclude` in `*.rst`.

File: docs/usage/help.rst.inc (hunk 1 above)
Comment (fantasya-pbem) [reply docs/usage/help.rst.inc:38]: ok

File: src/borg/archiver.py [hunk 3]
```diff
@@ -2489,15 +2489,12 @@ def do_break_lock(self, args, repository):
...
-        This means that e.g. ``borg create /path/to/repo ../some/path`` will
-        store all files as `some/path/.../file.ext` and ``borg create
-        /path/to/repo /home/user`` will store all files as
-        `home/user/.../file.ext`.
+        If you give `/absolute/` as root, the paths going into the matcher will
+        look relative like `absolute/.../file.ext`, because file paths in Borg
+        archives are always stored normalized and relative. This means that e.g.
+        ``borg create /path/to/repo ../some/path`` will store all files as
+        `some/path/.../file.ext` and ``borg create /path/to/repo /home/user``
+        will store all files as `home/user/.../file.ext`.
```
Comment (fantasya-pbem) [inline src/borg/archiver.py:2497]: Combined & shortened these two paragraphs.

Comment (ThomasWaldmann) [reply src/borg/archiver.py:2497]: good, thanks!

**Review by ThomasWaldmann — APPROVED** [review]
LGTM. Are you finished with this?

**Comment by fantasya-pbem** [comment]
Yes, and I have prepared a backport to 1.2-maint, will make a PR now.

**Comment by ThomasWaldmann** [comment]
Thanks!

**Pull request merged**